### ISSUE TITLE
Create Physics on the backend

### ISSUE THREAD

**Issue by AaronLieb** [AaronLieb/VideoGameHackingWorkshop#12]
Class Entity
- id
 - block:
 - onTick
 - position
 - velocity
 - acceleration

Session 
 - Level
    + Map
    + List<Entity>
    + Engine

**Comment by diamondburned** [comment]
Since we're adding entities, and each entities have a lot of their own information, instead of trying to encode the entities' properties over the wire, should we just make `Level` be shared between the server and the client?

This does mean the server and the client will both run the same physics engine. However, some desync may happen, and we might not have a way to deal with that. A way to solve this is to keep the physics engine strictly stateless and consider everything from the server to be the source of truth for the physics state. We still have some state in the Level, though. We would have to figure out a way to serialize and deserialize every single state within an entity.

An alternative solution that is much simpler is to omit all physics from the client and instead resolve all physics exclusively on the server, only serializing new positions back to the client.

Pros:
- Far easier to handle, one source of truth
- Much less information to serialize: instead of serializing all physical properties of an object, we only serialize positions

Cons:
- A tick of 15Hz (or every 4ms) means that there is roughly 4ms of latency on every user interaction. A 4ms delay is insignificant for such a game, but it's still worth noticing.
- We need to lerp (linear interpolate) the entities when they're moving, otherwise they'd be teleporting across the screen when moving. Lerping may require the client to estimate the object velocity and predict that, but it's nowhere near a full-fledged physics engine.

**Comment by AaronLieb** [comment]
I totally agree with the potential cons, but I think it would definitely be worth it to implement the server side only physics engine first and then see if there is any substantial problems with it before putting another engine on the client.

**Comment by diamondburned** [comment]
(Conversation continued in a private Google Docs.)
